feature-add-meal/src/test/java/com/kinandcarta/lib/add/meal/AddMealViewModelTest.kt

```diff
@@ -0,0 +1,118 @@
+ */
+
+class AddMealViewModelTest {
+
+    private val mainThreadSurrogate = newSingleThreadContext("UI thread")
+
+    var subject: AddMealViewModel = AddMealViewModel()
+
+    @Before
+    fun setUp() {
+        Dispatchers.setMain(mainThreadSurrogate)
+    }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain() // reset main dispatcher to the original Main dispatcher
+        mainThreadSurrogate.close()
+    }
+
+    @Test
+    fun test_whenEditTitle_then_updateMealFormTitle() {
+        val expected = "New Title"
+        subject.onEditTitle(expected)
+
+        assertEquals(expected, subject.meal.title)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+
+    @Test
+    fun test_whenEditAdditionalInformation_then_updateMealFormAadditionalInformation() {
+        val expected = "New Additional Information"
+        subject.onEditAdditionalInformation(expected)
+
+        assertEquals(expected, subject.meal.additionalInformation)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+    @Test
+    fun test_whenEditQuantity_then_updateQuantity() {
+        val expected = "42"
+        subject.onEditQuantity(expected)
+
+        assertEquals(expected, subject.meal.quantity)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+    @Test
+    fun test_whenEditTemperature_then_updateTemperature() {
+        val expected = false
+        subject.onEditTemperature(false)
+
+        assertEquals(expected, subject.meal.isHot)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+    @Test
+    fun test_whenEditAddress_then_updateAddress() {
+        val expected = "K+C Create Europe, 71 Collier Street, London N1 9BE"
+        subject.onEditAddress(expected)
+
+        assertEquals(expected, subject.meal.address)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+    @Test
+    fun test_whenOnEditAvailableFrom_then_updateAvailableFrom() {
+        val expected = LocalDateTime.of(2020, Month.DECEMBER, 1, 9, 42)
+        subject.onEditAvailableFrom(expected)
+
+        assertEquals(expected, subject.meal.availableFrom)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+    @Test
+    fun test_whenOnEditUseBy_then_updateUseBy() {
+        val expected = LocalDateTime.of(2020, Month.DECEMBER, 1, 9, 42)
+        subject.onEditUseBy(expected)
+
+        assertEquals(expected, subject.meal.useBy)
+        assertEquals(AddMealViewModel.State.Initial, subject.state)
+    }
+
+    @Test
+    fun test_whenSubmit_then_updateStateToLoading() {
+        runBlocking {
+            launch(Dispatchers.Main) {
+                assertEquals(AddMealViewModel.State.Initial, subject.state)
+                subject.onSubmit()
+                assertEquals(AddMealViewModel.State.Loading, subject.state)
+            }
+        }
+    }
+}
```

Comment:
Can't test the mealSDK. I found it very hard to mock yesterday as Mockito say it's a final class. 🤔

Reply:
I find mockk a lot nicer than Mockito nowadays. Abstracting the SDK behind an interface where you define only the methods you need would also make testing a lot easier. In this context, the meals SDK is actually a repository for you so you could hide it behind a `MealsRepository` interface.

Also - ViewModels can be written as unit test without the need of running them on an emulator. Let me know if you need some help.